binder/postBuild

```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+# Install the JupyterLab dask-labextension
+jupyter labextension install dask-labextension
+
+# Replace DASK_DASHBOARD_URL with the proxy location
+sed -i -e "s|DASK_DASHBOARD_URL|/user/${JUPYTERHUB_USER}/proxy/8787|g" binder/jupyterlab-workspace.json
+
+# Copy into the workspaces directory
+# TODO: what is the right filename to specify?
+# cp jupyterlab-workspace.json /home/jovyan/.jupyter/lab/workspaces 
```

Comment:
@ian-r-rose this file in particular could use a glance if you have time

Reply:
Some questioons:

1.  When I run jupyter-repo2docker on this repo JLab doesn't seem to be easily available (navigating to /lab doesn't work)
2.  When I uncomment this line, I get the following error `cp: cannot create regular file '/home/jovyan/.jupyter/lab/workspaces': No such file or directory`
3.  I don't know how to name this file in the workspaces directory to make sure that it gets picked up

Reply:
There is some name mangling that goes on for the workspace file (the precise reasons escape me right now, but it has something to do with avoiding file collisions). The best way to do this right now would be:
1. Go the url `/lab/workspaces/myworkspacename` in your JLab session
1. Make the layout you want in the session.
1. Go to your `~`/.jupyter/lab/workspaces` directory. Despite the mangling, you should be able to identify the file corresponding to your workspace.
1. Distribute this file *without changing the name* in the example repo.
1. In the postBuild script, copy this file to `/home/jovyan/.jupyter/lab/workspaces`.

It's ugly, but should work. Soon (in a matter of days, afaik) we should be able to merge a basic CLI into the jupyter lab application. So then the process would be something like
1. Repeat steps 1-2 above.
1. Run `jupyter lab workspace export myworkspacename > workspace.json`
1. Distribute `workspace.json` in the example repo.
1. In the postBuild script run `jupyter lab import myworspacename < workspace.json`

Reply:
I'm not sure why JupyterLab shouldn't be available, I'll give the branch a shot on mybinder.org

Reply:
Note that my description above should still work for the default, unnamed workspace, which is accessed under the `/lab` endpoint.

Reply:
As for your copying question: it's possible that JupyterLab doesn't create the workspaces directory until the first time it is used. I would expect this to be fixed once the CLI lands. In the meantime, maybe a `mkdir /home/jovyan/.jupyter/lab/workspaces` would do the trick?

Reply:
It looks like the JupyterLab server extension is disabled in this build. It should be enabled automatically, so I am not sure what the problem is.

It can be enabled by running `jupyter serverextension enable jupyterlab` in the `postBuild` script, but that really shouldn't be necessary.

Reply:
The base image used by `repo2docker` already includes JupyterLab (a few point releases behind), and that works as-is. There appears to be some weird interaction going on here that is *disabling* it. I'm not sure what that could be -- perhapse @yuvipanda has some insight?

Reply:
Maybe explicitly adding jupyterlab as a dependency without a version is causing problems? I'm never sure when conda is involved, maybe @minrk would know?

Reply:
In general, labextensions have to pretty strictly pin the version jupyterlab due to how it handles things.

Having unpinned dependencies in environment.yml means it will always install the latest version, regardless of what's in the env (unlike pip, which only consults the index for packages not already satisfied). [conda 4.6](https://github.com/conda/conda/pull/7291) should allow repo2docker to avoid updating packages already in the env, but the compatibility pinning issue remains.

Reply:
So if I don't include `jupyterlab` in my environment.yaml it appears that the current version is too old for `dask-labextension`

```
> /usr/bin/npm pack dask-labextension
dask-labextension-0.1.1.tgz

"dask-labextension@0.1.1" is not compatible with the current JupyterLab
Conflicting Dependencies:
JupyterLab              Extension            Package
>=0.13.1-0 <0.14.0-0    >=0.18.0-0 <0.19.0-0 @jupyterlab/application
>=0.13.1-0 <0.14.0-0    >=0.18.0-0 <0.19.0-0 @jupyterlab/apputils
>=0.13.0-0 <0.14.0-0    >=0.18.4-0 <0.19.0-0 @jupyterlab/console
>=0.13.0-0 <0.14.0-0    >=0.18.4-0 <0.19.0-0 @jupyterlab/notebook

cp: cannot create regular file '/home/jovyan/.jupyter/lab/workspaces': No such file or directory
Removing intermediate container 719799b65c64
```

If I include a pinned version of JupyterLab, like  `jupyterlab=0.34.7` then a couple of unfortunate things occur

```
> /usr/bin/npm pack dask-labextension
dask-labextension-0.1.1.tgz
> node /srv/conda/lib/python3.6/site-packages/jupyterlab/staging/yarn.js install
yarn install v1.9.4
info No lockfile found.
[1/4] Resolving packages...
warning css-loader > cssnano > autoprefixer > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
warning css-loader > cssnano > postcss-merge-rules > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
warning css-loader > cssnano > postcss-merge-rules > caniuse-api > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
[2/4] Fetching packages...
error webpack@4.12.2: The engine "node" is incompatible with this module. Expected version ">=6.11.5".
error Found incompatible module
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
> node /srv/conda/lib/python3.6/site-packages/jupyterlab/staging/yarn.js run build
yarn run v1.9.4
$ webpack
/bin/sh: 1: webpack: not found
error Command failed with exit code 127.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
cp: cannot create regular file '/home/jovyan/.jupyter/lab/workspaces': No such file or directory
Removing intermediate container 648853c6b798
The command '/bin/sh -c ./binder/postBuild' returned a non-zero code: 1
```

Reply:
The problem appears to be with the `default` channel. I don't know what JupyterLab they are working with, but if I remove that and work only with `conda-forge`, things behave much better.

Reply:
Still running into issues on my end

```diff
a/binder/environment.yml b/binder/environment.yml
index 46b5819..d53f8d1 100644
--- a/binder/environment.yml
+++ b/binder/environment.yml
@@ -1,12 +1,12 @@
 channels:
-  - defaults
   - conda-forge
+  - defaults
 dependencies:
   - bokeh=0.13
   - dask==0.19.0
   - dask-ml==0.9.0
   - distributed==1.23.0
-  - jupyterlab
+  - jupyterlab==0.34.7
   - numpy
   - pandas
   - pyarrow==0.10.0
```

```
Step 30/30 : RUN ./binder/postBuild
 ---> Running in ba3b020c6e2b
> /usr/bin/npm pack dask-labextension
dask-labextension-0.1.1.tgz
> node /srv/conda/lib/python3.6/site-packages/jupyterlab/staging/yarn.js install
yarn install v1.9.4
info No lockfile found.
[1/4] Resolving packages...
warning css-loader > cssnano > autoprefixer > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
warning css-loader > cssnano > postcss-merge-rules > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
warning css-loader > cssnano > postcss-merge-rules > caniuse-api > browserslist@1.7.7: Browserslist 2 could fail on reading Browserslist >3.0 config used in other tools.
[2/4] Fetching packages...
error webpack@4.12.2: The engine "node" is incompatible with this module. Expected version ">=6.11.5".
error Found incompatible module
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
> node /srv/conda/lib/python3.6/site-packages/jupyterlab/staging/yarn.js run build
yarn run v1.9.4
$ webpack
/bin/sh: 1: webpack: not found
error Command failed with exit code 127.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
cp: cannot create regular file '/home/jovyan/.jupyter/lab/workspaces': No such file or directory
Removing intermediate container ba3b020c6e2b
```

Reply:
Node is missing in the dependencies, which is required to install extensions. I'm going to make a PR against yours in a few.